File: ObjectDatabase/Api/src/main/java/org/elkoserver/objdb/ObjDBBase.java
package org.elkoserver.objdb;

import com.grack.nanojson.JsonParserException;
import org.elkoserver.foundation.json.ObjectDecoder;
import org.elkoserver.json.JsonArray;
import org.elkoserver.json.JsonObject;
import org.elkoserver.objdb.store.ObjectDesc;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

import java.lang.reflect.Array;
import java.time.Clock;
import java.util.*;
import java.util.function.Consumer;

import static org.elkoserver.json.JsonParsing.jsonObjectFromString;

/**
 * Base class for both local and remote concrete implementations of the ObjDB
 * interface.
 */
abstract class ObjDBBase implements ObjDB {
    /** Table mapping JSON object type tags to Java classes. */
    private final Map<String, Class<?>> myClasses = new HashMap<>();

    /** Application trace object for logging. */
    protected final Trace tr;
    private final TraceFactory traceFactory;
    private final Clock clock;

    /**
     * Constructor.
     */
    ObjDBBase(Trace appTrace, TraceFactory traceFactory, Clock clock) {
        tr = appTrace;
        this.traceFactory = traceFactory;
        this.clock = clock;
    }

    /**
     * Inform the object database about a mapping from a JSON object type tag
     * string to a Java class.
     *
     * @param tag  The JSON object type tag string.
     * @param type  The class that 'tag' labels.
     */
    public void addClass(String tag, Class<?> type) {
        myClasses.put(tag, type);
    }

    /**
     * Convert a parsed JSON object description into the object it describes.
     *
     * @param jsonObj  The object being decoded.
     *
     * @return the object described by 'jsonObj'.
     */
    Object decodeJSONObject(JsonObject jsonObj) {
        Object result = null;
        String typeTag = jsonObj.getString("type", null);
        if (typeTag != null) {
            Class<?> type = myClasses.get(typeTag);
            if (type != null) {
                result = ObjectDecoder.decode(type, jsonObj, this, traceFactory, clock);
            } else {
                tr.errorm("no class for type tag '" + typeTag + "'");
            }
        }
        return result;
    }

    /**
     * Decode a collection of JSON strings into an object.
     *
     * @param ref  Reference string for the object to be decoded.
     * @param results  A collection of object descriptors; one of these will be
     *    the object named by 'ref', others will be that object's contents.
     *
     * @return the decoded object, or null if it could not be decoded according
     *    to the specified parameters.
     */
    Object decodeObject(String ref, ObjectDesc[] results) {
        String objStr = null;
        for (ObjectDesc result : results) {
            if (ref.equals(result.ref())) {
                objStr = result.obj();
                break;
            }
        }
        if (objStr == null) {
            tr.errorm("no object retrieved from ODB for ref " + ref);
            return null;
        } else {
            try {
                JsonObject jsonObj = jsonObjectFromString(objStr);
                insertContents(jsonObj, results);
                return decodeJSONObject(jsonObj);
            } catch (JsonParserException e) {
                tr.errorm("object store syntax error getting " + ref + ": " +
                          e.getMessage());
                return null;
            }
        }
    }

    /**
     * Given an object reference, obtain the object(s) it refers to, from among
     * the collection of objects retrieved by the store.  If the reference
     * value is a string, the output is the object referenced by that string.
     * If the reference value is a JSON array of strings, the output value is
     * an array of the objects referenced by those strings.  Otherwise, the
     * result is null.
     *
     * In the case of an array result, if all of the elments are of a common
     * type, then the result will be an array of that type.  Otherwise, the
     * result will be an array of Object.
     *
     * @param refValue  The value(s) of the property before dereferencing.
     * @param objs  The objects returned by the store.
     *
     * @return  The value(s) of the property after dereferencing.
     */
    private Object dereferenceValue(Object refValue, ObjectDesc[] objs) {
        Object result = null;
        if (refValue instanceof JsonArray) {
            Iterator<Object> refs = ((JsonArray) refValue).iterator();
            Object[] contents = new Object[((JsonArray) refValue).size()];
            Class<?> resultClass = null;
            for (int i = 0; i < contents.length; ++i) {
                Object ref = refs.next();
                if (ref instanceof String) {
                    contents[i] = decodeObject((String) ref, objs);
                    if (contents[i] != null) {
                        Class<?> elemClass = contents[i].getClass();
                        if (resultClass == null) {
                            resultClass = elemClass;
                        } else if (elemClass.isAssignableFrom(resultClass)) {
                            resultClass = elemClass;
                        } else if (!resultClass.isAssignableFrom(elemClass)) {
                            resultClass = Object.class;
                        }
                    }
                } else {
                    contents[i] = null;
                }
            }
            result = contents;
            if (resultClass != Object.class) {
                result = Array.newInstance(resultClass, contents.length);
                System.arraycopy(contents, 0, result, 0, contents.length);
            }
        } else if (refValue instanceof String) {
            result = decodeObject((String) refValue, objs);
        }
        return result;
    }

    /**
     * Replace the properties of a JsonObject that describe the object's
     * contents with the contents objects themselves, as retrieved by the
     * store.
     *
     * Any property whose name begins with "ref$" is treated as an object
     * reference.  It is removed from the object and replaced with a new
     * property whose name has the "ref$" prefix stripped off and whose value
     * is the object or objects referenced.
     *
     * @param obj  The JsonObject whose contents are to be inserted.
     * @param results  The results returned by the store.
     */
    private void insertContents(JsonObject obj, ObjectDesc[] results) {
        List<Map.Entry<String, Object>> contentsProps = null;
        Iterator<Map.Entry<String, Object>> iter =
            obj.entrySet().iterator();
        while (iter.hasNext()) {
            Map.Entry<String, Object> entry = iter.next();
            String propName = entry.getKey();
            if (propName.startsWith("ref$")) {
                iter.remove();
                if (contentsProps == null) {
                    contentsProps =
                            new LinkedList<>();
                }
                contentsProps.add(entry);
            }
        }
        if (contentsProps != null) {
            for (Map.Entry<String, Object> entry : contentsProps) {
                Object prop = dereferenceValue(entry.getValue(), results);
                if (prop != null) {
                    obj.put(entry.getKey().substring(4), prop);
                }
            }
        }
    }

    /**
     * Load one or more named class descriptor objects.
     *
     * @param classDescRefs  A comma separated list of class descriptor object
     *    names.
     */
    void loadClassDesc(String classDescRefs) {
        addClass("classes", ClassDesc.class);
        addClass("class", ClassTagDesc.class);
        getObject("classes", null, new ClassDescReceiver("classes"));
        if (classDescRefs != null) {
            StringTokenizer tags = new StringTokenizer(classDescRefs, " ,;:");
            while (tags.hasMoreTokens()) {
                String tag = tags.nextToken();
                getObject(tag, null, new ClassDescReceiver(tag));
            }
        }
    }

    private class ClassDescReceiver implements Consumer<Object> {
        String myTag;
        ClassDescReceiver(String tag) {
            myTag = tag;
        }
        public void accept(Object obj) {
            ClassDesc classes = (ClassDesc) obj;
            if (classes != null) {
                tr.eventi("loading classDesc '" + myTag + "'");
                classes.useInODB(ObjDBBase.this, tr);
            } else {
                tr.errorm("unable to load classDesc '" + myTag + "'");
            }
        }
    }

    /**
     * Get the class associated with a given JSON type tag string.
     *
     * @param baseType  Base class from which result class must be derived.
     * @param typeName  JSON type tag identifying the desired class.
     *
     * @return a class named by 'typeName' suitable for assignment to a
     *    method or constructor parameter of class 'baseType'.
     */
    public Class<?> resolveType(Class<?> baseType, String typeName) {
        Class<?> result = myClasses.get(typeName);
        if (result != null && !baseType.isAssignableFrom(result)) {
            result = null;
        }
        return result;
    }
}

File: ObjectDatabase/Api/src/main/kotlin/org/elkoserver/objdb/ObjDbBase.kt
package org.elkoserver.objdb

import com.grack.nanojson.JsonParserException
import org.elkoserver.foundation.json.JsonToObjectDeserializer
import org.elkoserver.json.JsonArray
import org.elkoserver.json.JsonObject
import org.elkoserver.json.JsonParsing.jsonObjectFromString
import org.elkoserver.objdb.store.ObjectDesc
import org.elkoserver.util.trace.slf4j.Gorgel
import java.util.HashMap
import java.util.LinkedList
import java.util.StringTokenizer
import java.util.function.Consumer

/**
 * Base class for both local and remote concrete implementations of the ObjDb
 * interface.
 */
abstract class ObjDbBase(
        protected val gorgel: Gorgel,
        private val jsonToObjectDeserializer: JsonToObjectDeserializer) : ObjDb {
    /** Table mapping JSON object type tags to Java classes.  */
    private val myClasses: MutableMap<String, Class<*>> = HashMap()

    /**
     * Inform the object database about a mapping from a JSON object type tag
     * string to a Java class.
     *
     * @param tag  The JSON object type tag string.
     * @param type  The class that 'tag' labels.
     */
    override fun addClass(tag: String, type: Class<*>) {
        myClasses[tag] = type
    }

    /**
     * Convert a parsed JSON object description into the object it describes.
     *
     * @param jsonObj  The object being decoded.
     *
     * @return the object described by 'jsonObj'.
     */
    fun decodeJSONObject(jsonObj: JsonObject): Any? {
        var result: Any? = null
        val typeTag = jsonObj.getString<String?>("type", null)
        if (typeTag != null) {
            val type = myClasses[typeTag]
            if (type != null) {
                result = jsonToObjectDeserializer.decode(type, jsonObj, this)
            } else {
                gorgel.error("no class for type tag '$typeTag'")
            }
        }
        return result
    }

    /**
     * Decode a collection of JSON strings into an object.
     *
     * @param ref  Reference string for the object to be decoded.
     * @param results  A collection of object descriptors; one of these will be
     * the object named by 'ref', others will be that object's contents.
     *
     * @return the decoded object, or null if it could not be decoded according
     * to the specified parameters.
     */
    fun decodeObject(ref: String, results: Array<ObjectDesc>): Any? {
        val objStr: String? = results
                .firstOrNull { ref == it.ref }
                ?.obj
        return if (objStr == null) {
            gorgel.error("no object retrieved from ObjDb for ref $ref")
            null
        } else {
            try {
                val jsonObj = jsonObjectFromString(objStr)!!
                insertContents(jsonObj, results)
                decodeJSONObject(jsonObj)
            } catch (e: JsonParserException) {
                gorgel.error("object store syntax error getting $ref: ${e.message}")
                null
            }
        }
    }

    /**
     * Given an object reference, obtain the object(s) it refers to, from among
     * the collection of objects retrieved by the store.  If the reference
     * value is a string, the output is the object referenced by that string.
     * If the reference value is a JSON array of strings, the output value is
     * an array of the objects referenced by those strings.  Otherwise, the
     * result is null.
     *
     * In the case of an array result, if all of the elments are of a common
     * type, then the result will be an array of that type.  Otherwise, the
     * result will be an array of Object.
     *
     * @param refValue  The value(s) of the property before dereferencing.
     * @param objs  The objects returned by the store.
     *
     * @return  The value(s) of the property after dereferencing.
     */
    private fun dereferenceValue(refValue: Any, objs: Array<ObjectDesc>): Any? {
        var result: Any? = null
        if (refValue is JsonArray) {
            val refs = refValue.iterator()
            val contents = arrayOfNulls<Any>(refValue.size())
            var resultClass: Class<*>? = null
            for (i in contents.indices) {
                val ref = refs.next()
                if (ref is String) {
                    val decodedObject = decodeObject(ref, objs)
                    contents[i] = decodedObject
                    if (decodedObject != null) {
                        val elemClass: Class<*> = decodedObject.javaClass
                        if (resultClass == null) {
                            resultClass = elemClass
                        } else if (elemClass.isAssignableFrom(resultClass)) {
                            resultClass = elemClass
                        } else if (!resultClass.isAssignableFrom(elemClass)) {
                            resultClass = Any::class.java
                        }
                    }
                } else {
                    contents[i] = null
                }
            }
            result = contents
            if (resultClass != Any::class.java) {
                result = java.lang.reflect.Array.newInstance(resultClass, contents.size)
                System.arraycopy(contents, 0, result, 0, contents.size)
            }
        } else if (refValue is String) {
            result = decodeObject(refValue, objs)
        }
        return result
    }

    /**
     * Replace the properties of a JsonObject that describe the object's
     * contents with the contents objects themselves, as retrieved by the
     * store.
     *
     * Any property whose name begins with "ref$" is treated as an object
     * reference.  It is removed from the object and replaced with a new
     * property whose name has the "ref$" prefix stripped off and whose value
     * is the object or objects referenced.
     *
     * @param obj  The JsonObject whose contents are to be inserted.
     * @param results  The results returned by the store.
     */
    private fun insertContents(obj: JsonObject, results: Array<ObjectDesc>) {
        var contentsProps: MutableList<Map.Entry<String, Any>>? = null
        val iter: MutableIterator<Map.Entry<String, Any>> = obj.entrySet().iterator()
        while (iter.hasNext()) {
            val entry = iter.next()
            val propName = entry.key
            if (propName.startsWith("ref$")) {
                iter.remove()
                if (contentsProps == null) {
                    contentsProps = LinkedList()
                }
                contentsProps.add(entry)
            }
        }
        if (contentsProps != null) {
            for ((key, value) in contentsProps) {
                val prop = dereferenceValue(value, results)
                if (prop != null) {
                    obj.put(key.substring(4), prop)
                }
            }
        }
    }

    /**
     * Load one or more named class descriptor objects.
     *
     * @param classDescRefs  A comma separated list of class descriptor object
     * names.
     */
    fun loadClassDesc(classDescRefs: String?) {
        addClass("classes", ClassDesc::class.java)
        addClass("class", ClassTagDesc::class.java)
        getObject("classes", null, ClassDescReceiver("classes"))
        if (classDescRefs != null) {
            val tags = StringTokenizer(classDescRefs, " ,;:")
            while (tags.hasMoreTokens()) {
                val tag = tags.nextToken()
                getObject(tag, null, ClassDescReceiver(tag))
            }
        }
    }

    private inner class ClassDescReceiver(var myTag: String) : Consumer<Any?> {
        override fun accept(obj: Any?) {
            val classes = obj as ClassDesc?
            if (classes != null) {
                gorgel.i?.run { info("loading classDesc '$myTag'") }
                classes.useInObjDb(this@ObjDbBase, gorgel)
            } else {
                gorgel.error("unable to load classDesc '$myTag'")
            }
        }

    }

    /**
     * Get the class associated with a given JSON type tag string.
     *
     * @param baseType  Base class from which result class must be derived.
     * @param typeName  JSON type tag identifying the desired class.
     *
     * @return a class named by 'typeName' suitable for assignment to a
     * method or constructor parameter of class 'baseType'.
     */
    override fun resolveType(baseType: Class<*>, typeName: String): Class<*>? {
        var result = myClasses[typeName]
        if (result != null && !baseType.isAssignableFrom(result)) {
            result = null
        }
        return result
    }

}